cns/restserver/ipam.go

```diff
@@ -478,101 +634,194 @@ func (service *HTTPRestService) MarkExistingIPsAsPendingRelease(pendingIPIDs []s
+				if len(assignedMap) == len(desiredIPAddresses) {
+					return podIPInfo, err
 				}
-			default:
-				return podIpInfo, errors.Errorf("[AllocateDesiredIPConfig] Desired IP is not available %+v", ipConfig)
+			} else {
+				logger.Errorf("[AssignDesiredIPConfigs] Desired IP is already assigned %+v, requested for pod %+v", ipConfig, podInfo)
+				//nolint:goerr113 // return error
+				return podIPInfo, fmt.Errorf("IP already assigned to another pod")
+			}
+		case types.Available, types.PendingProgramming:
+			// This race can happen during restart, where CNS state is lost and thus we have lost the NC programmed version
+			ncMap[ipConfig.NCID] = ipConfig
+		default:
+			logger.Errorf("[AssignDesiredIPConfigs] Desired IP is not available %+v", ipConfig)
+			//nolint:goerr113 // return error
+			return podIPInfo, fmt.Errorf("IO not available")
+		}
+		if len(ncMap) == len(desiredIPAddresses) {
+			break
+		}
+	}
+
+	podIPctr := 0
+	// assigns all IPs in the map to the pod
+	for id := range ncMap {
+
+		// As part of reconcile, we mark IPs as Assigned which are already assigned to Pods (listed from APIServer)
+		if err := service.assignIPConfig(ncMap[id], podInfo); err != nil {
+			logger.Errorf(err.Error())
+			break
+		}
+
+		if err := service.populateIPConfigInfoUntransacted(ncMap[id], &podIPInfo[podIPctr]); err != nil {
+			logger.Errorf(err.Error())
+			break
+		}
+		podIPctr++
+		// Checks if we no longer have any IPs to be added and returns if we don't
+		if podIPctr == len(service.state.ContainerStatus) {
+			return podIPInfo, nil
+		}
+	}
+
+	// if we were able to get at least one IP but not all of the desired IPs
+	if len(ncMap) > 0 {
+		logger.Printf("[AssignDesiredIPConfigs] Failed to retrieve all desired IPs. Releasing all IPs that were found")
+		for _, ipState := range ncMap { //nolint:gocritic // ignore copy
+			_, err := service.unassignIPConfig(ipState, podInfo)
+			if err != nil {
+				return podIPInfo, fmt.Errorf("[AssignDesiredIPConfigs] failed to mark IPConfig [%+v] back to Available. err: %w", ipState, err)
 			}
-			err := service.populateIPConfigInfoUntransacted(ipConfig, &podIpInfo)
-			return podIpInfo, err
 		}
 	}
-	return podIpInfo, fmt.Errorf("Requested IP not found in pool")
+	//nolint:goerr113 // return error
+	return podIPInfo, fmt.Errorf("not all requested ips were found/available in the pool")
 }
 
-func (service *HTTPRestService) AssignAnyAvailableIPConfig(podInfo cns.PodInfo) (cns.PodIpInfo, error) {
+// Assigns an IP from each NC on the NNC
+func (service *HTTPRestService) AssignAvailableIPConfigs(podInfo cns.PodInfo) ([]cns.PodIpInfo, error) {
 	service.Lock()
 	defer service.Unlock()
+	// Creates a slice of PodIpInfo with the size of number of NCs
+	podIPInfo := make([]cns.PodIpInfo, len(service.state.ContainerStatus))
+	// This map is used to store whether or not we have found an available IP from an NC when looping through the pool
+	ncMap := make(map[string]cns.IPConfigurationStatus)
```

Comment:
can you change the name of the map to reflect what it holds? ncMap doesn't indicate what it holds